library/src/main/java/org/apache/apex/malhar/lib/window/impl/AbstractWindowedJoinOperator.java

```diff
@@ -0,0 +1,125 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.apex.malhar.lib.window.impl;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.apache.apex.malhar.lib.window.ControlTuple;
+import org.apache.apex.malhar.lib.window.JoinAccumulation;
+import org.apache.apex.malhar.lib.window.JoinWindowedOperator;
+import org.apache.apex.malhar.lib.window.Tuple;
+import org.apache.apex.malhar.lib.window.WindowedStorage;
+
+import com.google.common.base.Function;
+import com.datatorrent.api.DefaultInputPort;
+import com.datatorrent.api.annotation.InputPortFieldAnnotation;
+
+
+/**
+ * Abstract Windowed Join Operator.
+ */
+public abstract class AbstractWindowedJoinOperator<InputT1, InputT2, OutputT, DataStorageT extends WindowedStorage,
+    RetractionStorageT extends WindowedStorage, AccumulationT extends
+    JoinAccumulation>
+    extends AbstractWindowedOperator<InputT1, OutputT, DataStorageT, RetractionStorageT, AccumulationT>
+    implements JoinWindowedOperator<InputT1, InputT2>
+{
+
+  private static final transient Logger LOG = LoggerFactory.getLogger(AbstractWindowedJoinOperator.class);
+  private Function<InputT2, Long> timestampExtractor2;
+
+  private long latestWatermark1 = -1;  // latest watermark from stream 1
+  private long latestWatermark2 = -1;  // latest watermark from stream 2
+
+  public final transient DefaultInputPort<Tuple<InputT2>> input2 = new DefaultInputPort<Tuple<InputT2>>()
+  {
+    @Override
+    public void process(Tuple<InputT2> tuple)
+    {
+      processTuple2(tuple);
+    }
+  };
+
+  // TODO: This port should be removed when Apex Core has native support for custom control tuples
+  @InputPortFieldAnnotation(optional = true)
+  public final transient DefaultInputPort<ControlTuple> controlInput2 = new DefaultInputPort<ControlTuple>()
+  {
+    @Override
+    public void process(ControlTuple tuple)
+    {
+      if (tuple instanceof ControlTuple.Watermark) {
+        processWatermark2((ControlTuple.Watermark)tuple);
+      }
+    }
+  };
+
+  public void processTuple2(Tuple<InputT2> tuple)
+  {
+    long timestamp = extractTimestamp(tuple, timestampExtractor2);
+    if (isTooLate(timestamp)) {
+      dropTuple(tuple);
+    } else {
+      Tuple.WindowedTuple<InputT2> windowedTuple = getWindowedValueWithTimestamp(tuple, timestamp);
+      // do the accumulation
+      accumulateTuple2(windowedTuple);
+      processWindowState(windowedTuple);
+    }
+  }
+
+  public void setTimestampExtractor2(Function<InputT2, Long> timestampExtractor2)
+  {
+    this.timestampExtractor2 = timestampExtractor2;
+  }
+
+
+  @Override
+  public void processWatermark(ControlTuple.Watermark watermark)
+  {
+    latestWatermark1 = watermark.getTimestamp();
+    if (this.watermarkTimestamp < 0) {
+      this.watermarkTimestamp = watermark.getTimestamp();
```

Comment:
Also, please add a unit test for this case to make sure watermark will not be fired downstream until we have watermark from both input ports